blog-posts/typescript-ast-type-checker.md

```diff
@@ -0,0 +1,591 @@
+      ├── setSeconds: (sec: number, ms?: number) => number
+      ├── setUTCSeconds: (sec: number, ms?: number) => number
+      ├── setMinutes: (min: number, sec?: number, ms?: number) => number
+      ├── setUTCMinutes: (min: number, sec?: number, ms?: number) => number
+      ├── setHours: (hours: number, min?: number, sec?: number, ms?: number) => number
+      ├── setUTCHours: (hours: number, min?: number, sec?: number, ms?: number) => number
+      ├── setDate: (date: number) => number
+      ├── setUTCDate: (date: number) => number
+      ├── setMonth: (month: number, date?: number) => number
+      ├── setUTCMonth: (month: number, date?: number) => number
+      ├── setFullYear: (year: number, month?: number, date?: number) => number
+      ├── setUTCFullYear: (year: number, month?: number, date?: number) => number
+      ├── toUTCString: () => string
+      ├── toISOString: () => string
+      ├── toJSON: (key?: any) => string
+      ├── getVarDate: () => VarDate
+      ├── __@toPrimitive@755: { (hint: "default"): string; (hint: "string"): string; (hint: "number"): number; (hint: string): string | number; }
+    👉├── value3: Date
+    👉├── collectedValue: ImportantValue
+    ```
+</details>
+
+That's the "standard library" issue described earlier. The `Date` and `string` types causes this drama and we need to stop our processing before entering those types.
+
+### Exclude the standard types
+
+TypeScript gives us plenty of tools to do that. Here is a helper method `isTypeLocal` I have build for our use cases.
+
+```typescript
+function isTypeLocal(symbol: ts.Symbol) {
+  const sourceFile = symbol?.valueDeclaration?.getSourceFile();
+  const hasSource = !!sourceFile;
+  const isStandardLibrary = hasSource && program.isSourceFileDefaultLibrary(sourceFile!)
+  const isExternal = hasSource && program.isSourceFileFromExternalLibrary(sourceFile!);
+  const hasDeclaration = !!symbol?.declarations?.[0];
+
+  return !(isStandardLibrary || isExternal) && hasDeclaration;
+}
+```
+
+The method will detect if a given symbol belongs to a standard library (`Date`), to an external library (whatever you use from `node_modules`) and everything that doesn't have an actual declaration like primitive types (`string`, `number`).
+
+We will use that helper to prevent our recursion from branching into those unwanted types:
+
+```typescript
+if(isTypeLocal(propertySymbol)) {
+  // It's a type we have defined, so print it
+  // and then process its nested properties
+  console.group(`  └── ${property.name}: ${propertyTypeName}`)
+  processProperty(propertyType, node, level + 1)
+}else {
+  // It's not a local type, so print it but don't do anything further
+  console.log(`  ├── ${property.name}: ${propertyTypeName}`)
+}
+```
+
+The updated code can process the initial file but it's much more flexible. Let's process a much deeper nested type `Output` and watch the console.
+
+<details>
+    <summary>Updated file `file-with-types.ts`</summary>
```

Comment:
Not formatted. Also a candidate for a separate PR

<img width="434" alt="Screenshot 2021-06-21 at 08 09 24" src="https://user-images.githubusercontent.com/7814926/122714692-022f0600-d268-11eb-86f4-6cac095d08be.png">

Reply:
yes fixed with the correct indentation.